features/steps/thens/relations.py

```diff
@@ -131,6 +126,155 @@ def step_impl(context, related, relating, other_entity, condition):
+                    yield (err.InvalidValueError(False, inst, 'Name', name))  # A custom Pset_ prefixed attribute, e.g. Pset_Mywall
+
+            accepted_values = establish_accepted_pset_values(name, property_set_definitons)
+
+            if accepted_values:  # If not it's a custom Pset_ prefixed attribute, e.g. Pset_Mywall (no need for further Pset_ checks),
+
+                if 'IfcPropertySet must be assigned according to the property set definitions table' in context.step.name:
+                    try:
+                        relations = inst.PropertyDefinitionOf  # IFC2x3 https://standards.buildingsmart.org/IFC/RELEASE/IFC2x3/TC1/HTML/ifckernel/lexical/ifcpropertysetdefinition.htm
+                    except AttributeError: # IFC4-CHANGE Inverse attribute renamed from PropertyDefinitionOf with upward compatibility for file-based exchange.
+                        # https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcPropertySet.htm
+                        relations = inst.DefinesOccurrence
+                    except IndexError:  # IfcPropertySet not assigned to IfcObjects
+                        relations = []
+
+                    for relation in relations:
+                        related_objects = relation.RelatedObjects
+                        for obj in related_objects:
+
+                            if accepted_values['template_type'] and accepted_values['template_type'] in ['PSET_TYPEDRIVENONLY']:
+                                yield (err.InvalidPropertySetDefinition(False, inst=inst, object=obj, name=name, template_type_enum=accepted_values['template_type']))
+
+                            correct = [obj.is_a(accepted_object) for accepted_object in accepted_values['applicable_entities']]
+                            if not any(correct):
+                                yield (err.InvalidPropertySetDefinition(False, inst, obj, name, accepted_values['applicable_entities']))
+                            elif context.error_on_passed_rule:
+                                yield (err.RuleSuccessInst(True, inst))
+
+                    related_objects = inst.DefinesType
+                    for obj in related_objects:
+                        if accepted_values['template_type'] and accepted_values['template_type'] in ['PSET_OCCURRENCEDRIVEN', 'PSET_PERFORMANCEDRIVEN']:
+                            yield (err.InvalidPropertySetDefinition(False, inst=inst, object=obj, name=name, template_type_enum=accepted_values['template_type']))
+
+                        correct = [obj.is_a(accepted_object) for accepted_object in accepted_values['applicable_type_values']]
+                        if not any(correct):
+                            yield (err.InvalidPropertySetDefinition(False, inst, obj, name, accepted_values['applicable_type_values']))
+                        elif context.error_on_passed_rule:
+                            yield (err.RuleSuccessInst(True, inst))
```

Comment:
This monster block for all the rules is a bit funky. I think it's fine for now. My main point would be that it's not clear when the RuleSuccessInst is supposed to be emitted, but anyway that is going to have to be refactored with the other PR coming in.

I would probably prefer 5 differnet functions that make use of some utility function, but either way is fine.

Reply:
I agree. 

My plan of action:
 1. Merge this to main.
 2. Merge main into PR decorator branch. 
 3. Change PSE001 part in the decorator branch. 
 
 I think it's the cleanest way to tackle it.